Clockwise.Tests/CancellationTokenBudgetTests.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Globalization;
+using FluentAssertions;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+using static System.Environment;
+
+namespace Clockwise.Tests
+{
+    public class CancellationTokenBudgetTests : IDisposable
+    {
+        protected void StartClock(DateTimeOffset? now = null) => VirtualClock.Start(now);
+
+        public void Dispose() => Clock.Reset();
+
+        [Fact]
+        public void When_the_budget_is_created_then_the_start_time_is_captured()
+        {
+            StartClock();
+
+            var budget = new TimeBudget(5.Seconds());
+
+            budget.StartTime.Should().Be(Clock.Now());
+        }
+
+        [Fact]
+        public async Task CancellationTokenBudget_throws_an_informative_exception_after_it_is_cancelled()
+        {
+            StartClock();
+
+            var budget = new CancellationTokenBudget();
+            
+            await Clock.Current.Wait(1.Seconds());
+
+            budget.RecordEntry("one");
+
+            await Clock.Current.Wait(10.Seconds());
+
+            budget.Cancel();
+
+            Action action = () => budget.RecordEntryAndThrowIfBudgetExceeded("two");
```

Comment:
could we have a shorter method name and optional parameter throwIfBudgerEceeded as an alternative design?

Reply:
The reason I prefer separate methods for such different behaviors is that the intent isn't very clear when reading something like this:

```csharp
budget.RecordEntry(true);
```

Reply:
agreed